### ISSUE TITLE
data type VARCHAR is not compatible for mysql 5.7 when store Chinese character

### ISSUE THREAD

**Issue by vkingnew** [pingcap/tidb#9760]
# MySQL

```
>mysql> select version();
+------------+
| version()  |
+------------+
| 5.7.15-log |
+------------+
1 row in set (0.00 sec)
>create table t(id bigint ,remark varchar(255));
>insert into t(id,remark)values(1,REPEAT('汉',256));
ERROR 1406 (22001): Data too long for column 'remark' at row 1
>mysql> insert into t(id,remark)values(1,REPEAT('汉',255));
Query OK, 1 row affected (0.00 sec)

>mysql> insert into t(id,remark)values(1,REPEAT('汉',86));
Query OK, 1 row affected (0.00 sec)

>mysql> insert into t(id,remark)values(1,REPEAT('汉',85));
Query OK, 1 row affected (0.00 sec)
mysql> show create table t\G
*************************** 1. row ***************************
       Table: t
Create Table: CREATE TABLE `t` (
  `id` bigint(20) DEFAULT NULL,
  `remark` varchar(255) COLLATE utf8mb4_unicode_ci DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
1 row in set (0.00 sec)
```

#  TiDB:

```
> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v2.1.5-1-g0bd6b1b
Git Commit Hash: 0bd6b1b60816bf8a2022e657d7a472311548d82e
Git Branch: release-2.1
UTC Build Time: 2019-02-28 08:17:03
GoVersion: go version go1.11.2 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false
1 row in set (0.00 sec)
>mysql> insert into t(id,remark)values(1,REPEAT('汉',256));
ERROR 1406 (22001): Data too long for column 'remark' at row 1
mysql> insert into t(id,remark)values(1,REPEAT('汉',255));
ERROR 1406 (22001): Data too long for column 'remark' at row 1
mysql> insert into t(id,remark)values(1,REPEAT('汉',86));
ERROR 1406 (22001): Data too long for column 'remark' at row 1
mysql> insert into t(id,remark)values(1,REPEAT('汉',85));
Query OK, 1 row affected (0.01 sec)

>mysql> show create table t\G
*************************** 1. row ***************************
       Table: t
Create Table: CREATE TABLE `t` (
  `id` bigint(20) DEFAULT NULL,
  `remark` varchar(255) CHARSET UTF8MB4 COLLATE utf8mb4_bin DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=UTF8MB4 COLLATE=UTF8MB4_UNICODE_CI
1 row in set (0.00 sec)
```

# conclusion：
>varchar(255) In mysql can store max 255 Chinese character,how ever tidb just store 255/3=85 Chinese character,it's not compatile for mysql, when migration mysql to tidb,it's a hard work to do.

**Comment by vkingnew** [comment]
which version could be fixed this bug and when release ?

**Comment by morgo** [comment]
To summarize this problem, in MySQL the _n_ in `varchar(n)` is the maximum number of characters.  So `varchar(255)` could take up to 255*4 bytes in utf8mb4.  In TiDB, the _n_ is the maximum number of bytes.

I believe the SQL standard refers to _n_ being characters as well.  We should at least document this compatibility difference for now.

**Comment by eurekaka** [comment]
@jackysp PTAL

**Comment by vkingnew** [comment]
varchar(N) in MySQL is not SQL standard,just MySQL dialect.when migrate from MySQL to TiDB,we have to change many columns if not compatile for mysql.  @morgo @jackysp

**Comment by jackysp** [comment]
Hi @vkingnew , it is already fixed on the master branch, we'll release 3.0-beta.2 this week, which could include this bug fix.

**Comment by vkingnew** [comment]
this bug will be fixed in TiDB 2.1 branch？

**Comment by jackysp** [comment]
> this bug will be fixed in TiDB 2.1 branch？ 
Please confirm it @winkyao .
> And about when to release TiDB Release 3.0 GA?
In this June.

**Comment by vkingnew** [comment]
@winkyao  
>this bug will be fixed in TiDB 2.1 branch？

**Comment by morgo** [comment]
Just to confirm, I simplified the test case and verified in master that this is fixed.  The error code is also identical to MySQL:
```
DROP TABLE IF EXISTS t1;
CREATE TABLE t1 (id bigint, remark varchar(255));
INSERT INTO t1 (id,remark) VALUES (1,REPEAT('汉',256)); -- error in MySQL/TiDB
INSERT INTO t1 (id,remark) VALUES (1,REPEAT('汉',255));
INSERT INTO t1 (id,remark) VALUES (1,REPEAT('汉',86));

mysql> DROP TABLE IF EXISTS t1;
Query OK, 0 rows affected (0.00 sec)

mysql> CREATE TABLE t1 (id bigint, remark varchar(255));
Query OK, 0 rows affected (0.00 sec)

mysql> INSERT INTO t1 (id,remark) VALUES (1,REPEAT('汉',256)); -- error in MySQL/TiDB
ERROR 1406 (22001): Data too long for column 'remark' at row 1
mysql> INSERT INTO t1 (id,remark) VALUES (1,REPEAT('汉',255));
Query OK, 1 row affected (0.01 sec)

mysql> INSERT INTO t1 (id,remark) VALUES (1,REPEAT('汉',86));
Query OK, 1 row affected (0.00 sec)
```

**Comment by vkingnew** [comment]
if the columns define as VARCHAR(255),it can on store max 255 Chinese character,so you test cases are right .

**Comment by vkingnew** [comment]
this feature is fixed by TiDB version 2.1.7:
>mysql> DROP TABLE IF EXISTS t1;
Query OK, 0 rows affected (0.00 sec)
mysql> CREATE TABLE t1 (id bigint, remark varchar(255));
Query OK, 0 rows affected (0.08 sec)
mysql> INSERT INTO t1 (id,remark) VALUES (1,REPEAT('汉',255));
Query OK, 1 row affected (0.11 sec)
mysql> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v2.1.7
Git Commit Hash: f5b52cb9dc5a8a546121aa6ce1fc672bd567ca48
Git Branch: HEAD
UTC Build Time: 2019-03-28 07:54:41
GoVersion: go version go1.11.2 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false
1 row in set (0.00 sec)

**Comment by winkyao** [comment]
@vkingnew This is a compatibility issue, we will fix it. thanks
